plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php

```diff
@@ -649,4 +654,141 @@ private function print_hidden_form_fields(): void {
+
+		/**
+		 * Allows 3rd parties to modify order IDs about to be affected by a bulk action.
+		 *
+		 * @param array Array of order IDs.
+		 */
+		$ids = apply_filters( // phpcs:ignore WooCommerce.Commenting.CommentHooks.MissingSinceComment
+			'woocommerce_bulk_action_ids',
+			isset( $_REQUEST['order'] ) ? array_reverse( array_map( 'absint', $_REQUEST['order'] ) ) : array(),
+			$action,
+			'order'
+		);
+
+		if ( ! $ids ) {
+			wp_safe_redirect( $redirect_to );
+			exit;
+		}
+
+		$report_action = '';
+		$changed       = 0;
+
+		if ( 'remove_personal_data' === $action ) {
+			$report_action = 'removed_personal_data';
+			$changed       = $this->do_bulk_action_remove_personal_data( $ids );
+		} elseif ( false !== strpos( $action, 'mark_' ) ) {
+			$order_statuses = wc_get_order_statuses();
+			$new_status     = substr( $action, 5 );
+			$report_action  = 'marked_' . $new_status;
+
+			if ( isset( $order_statuses[ 'wc-' . $new_status ] ) ) {
+				$changed = $this->do_bulk_action_mark_orders( $ids, $new_status );
+			}
+		}
+
+		if ( $changed ) {
+			$redirect_to = add_query_arg(
+				array(
+					'bulk_action' => $report_action,
+					'changed'     => $changed,
+					'ids'         => implode( ',', $ids ),
+				),
+				$redirect_to
+			);
+		}
+
+		wp_safe_redirect( $redirect_to );
+		exit;
+	}
+
+	/**
+	 * Implements the "remove personal data" bulk action.
+	 *
+	 * @param array $order_ids The Order IDs.
+	 * @return int Number of orders modified.
+	 */
+	private function do_bulk_action_remove_personal_data( $order_ids ): int {
+		$changed = 0;
+
+		foreach ( $order_ids as $id ) {
+			$order = wc_get_order( $id );
+
+			if ( $order ) {
+				do_action( 'woocommerce_remove_order_personal_data', $order ); // phpcs:ignore WooCommerce.Commenting.CommentHooks.MissingHookComment
+				$changed++;
+			}
+		}
+
+		return $changed;
+	}
+
+	/**
+	 * Implements the "mark <status>" bulk action.
+	 *
+	 * @param array  $order_ids  The order IDs to change.
+	 * @param string $new_status The new order status.
+	 * @return int Number of orders modified.
+	 */
+	private function do_bulk_action_mark_orders( $order_ids, $new_status ): int {
+		$changed = 0;
+
+		// Initialize payment gateways in case order has hooked status transition actions.
+		WC()->payment_gateways();
+
+		foreach ( $order_ids as $id ) {
+			$order = wc_get_order( $id );
+			$order->update_status( $new_status, __( 'Order status changed by bulk edit:', 'woocommerce' ), true );
+			do_action( 'woocommerce_order_edit_status', $id, $new_status ); // phpcs:ignore WooCommerce.Commenting.CommentHooks.MissingHookComment
+			$changed++;
+		}
+
+		return $changed;
+	}
+
+	/**
+	 * Show confirmation message that order status changed for number of orders.
+	 */
+	public function bulk_action_notices() {
+		if ( empty( $_REQUEST['bulk_action'] ) ) {
+			return;
+		}
+
+		$order_statuses = wc_get_order_statuses();
+		$number         = isset( $_REQUEST['changed'] ) ? absint( $_REQUEST['changed'] ) : 0;
```

Comment:
Another non-essential tweak we could consider:

```suggestion
		$number         = absint( $_REQUEST['changed'] ?? 0 );
```